react-components/src/components/Architecture/OptionButton.tsx

```diff
@@ -0,0 +1,111 @@
+/*!
+ * Copyright 2023 Cognite AS
+ */
+
+import { useEffect, useMemo, useState, type ReactElement } from 'react';
+import { Button, Dropdown, Menu, Tooltip as CogsTooltip, type IconType } from '@cognite/cogs.js';
+import { useTranslation } from '../i18n/I18n';
+import { type BaseCommand } from '../../architecture/base/commands/BaseCommand';
+import { useRenderTarget } from '../RevealCanvas/ViewerContext';
+import { getDefaultCommand } from './CommandButton';
+import { BaseOptionCommand } from '../../architecture/base/commands/BaseOptionCommand';
+import styled from 'styled-components';
+import { getButtonType, getIcon, getTooltipPlacement } from './Utilities';
+
+export const OptionButton = ({
+  command,
+  isHorizontal = false
+}: {
+  command: BaseOptionCommand;
+  isHorizontal: boolean;
+}): ReactElement => {
+  const renderTarget = useRenderTarget();
+  const { t } = useTranslation();
+  const newCommand = useMemo<BaseCommand>(() => getDefaultCommand(command, renderTarget), []);
+
+  const [isOpen, setOpen] = useState<boolean>(false);
+  const [isEnabled, setEnabled] = useState<boolean>(true);
+  const [isVisible, setVisible] = useState<boolean>(true);
+  const [uniqueId, setUniqueId] = useState<number>(0);
+  const [icon, setIcon] = useState<IconType | undefined>(undefined);
+
+  useEffect(() => {
+    function update(command: BaseCommand): void {
+      setEnabled(command.isEnabled);
+      setVisible(command.isVisible);
+      setUniqueId(command.uniqueId);
+      setIcon(getIcon(command));
+    }
```

Comment:
Maybe is better to move this update function out of the useEffect and use the hook useCallback

Reply:
I don't know why or why not. I just do things that is working. Can you suggest by code it?

Reply:
if we define a function inside the `useEffect`, then the compiler will re create the same function every time when that `useEffect `is triggered and, in this case, with no reason.
So one way to avoid that is to or define this "regular" function outside of the `useEffect `or define it as a `useCallback `hook (outside the `useEffect `as well). `useCallback `will use the memoization mechanism for the `update `function.
So maybe something like this:

```typescript
   const update = useCallback((command: BaseCommand) => {
    setEnabled(command.isEnabled);
    setVisible(command.isVisible);
    setUniqueId(command.uniqueId);
    setIcon(getIcon(command));
  }, [setEnabled, setVisible, setUniqueId, setIcon]);
  

useEffect(() => {
    update(newCommand);
    newCommand.addEventListener(update);
    return () => {
      newCommand.removeEventListener(update);
    };
 }, [newCommand]);
```

Reply:
Ok, I see. This is now fixed